helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacityManager.java

```diff
@@ -0,0 +1,111 @@
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Manager class responsible for abstracting the management operations on the {@link WagedInstanceCapacity}
+ * cache such as rebuild and update operations on various scenarios.
+ */
+public class WagedInstanceCapacityManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(WagedInstanceCapacityManager.class);
+
+  private static class SingletonHelper {
+    private static final WagedInstanceCapacityManager INSTANCE = new WagedInstanceCapacityManager();
+  }
+
+  public static WagedInstanceCapacityManager getInstance() {
+    return SingletonHelper.INSTANCE;
+  }
+
+  public void processEvent(ClusterEvent event, CurrentStateOutput currentStateOutput) {
+    BaseControllerDataProvider cache = event.getAttribute(AttributeName.ControllerDataProvider.name());
+    final Map<String, Resource> resourceMap = event.getAttribute(AttributeName.RESOURCES.name());
+
+    if (shouldNoOp(cache, resourceMap, event)) {
+      return;
+    }
+
+    ResourceControllerDataProvider dataProvider = (ResourceControllerDataProvider) cache;
+    Map<String, Resource> wagedEnabledResourceMap = resourceMap.entrySet()
+        .stream()
+        .filter(entry -> WagedValidationUtil.isWagedEnabled(cache.getIdealState(entry.getKey())))
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    if (wagedEnabledResourceMap.isEmpty()) {
+      return;
+    }
+
+    // Phase 1: Rebuild Always
+    // DONE: we only need to compute when there are resource using Waged. We should
+    // do this as perf improvement in the future.
+    WagedInstanceCapacity capacityProvider = new WagedInstanceCapacity(dataProvider);
+    WagedResourceWeightsProvider weightProvider = new WagedResourceWeightsProvider(dataProvider);
+
+    // Process the currentState and update the available instance capacity.
+    capacityProvider.process(dataProvider, currentStateOutput, wagedEnabledResourceMap, weightProvider);
+    dataProvider.setWagedCapacityProviders(capacityProvider, weightProvider);
+  }
+
+  /**
+   * Function that checks whether we should return early, without any action on the capacity map or not.
+   * @param cache it is the cluster level cache for the resources.
+   * @param event the cluster event that is undergoing processing.
+   * @return true, of the condition evaluate to true and no action is needed, else false.
+   */
+  static boolean shouldNoOp(BaseControllerDataProvider cache, Map<String, Resource> resourceMap, ClusterEvent event) {
+    if (!(cache instanceof ResourceControllerDataProvider)) {
+      return true;
+    }
+
+    if (resourceMap == null || resourceMap.isEmpty()) {
+      return true;
+    }
+
+    ResourceControllerDataProvider dataProvider = (ResourceControllerDataProvider) cache;
+    if (Objects.isNull(dataProvider.getWagedInstanceCapacity())) {
+      return false;
+    }
+
+    switch (event.getEventType()) {
+      case TaskCurrentStateChange:
+      case MessageChange:
```

Comment:
MessageChange - won't this result in change in pendingMessages ?

Reply:
Looking at ClusterEventType, there are many events for which we can skip re-building cache. 
-  IdealStateChange,  CustomizedStateChange,  CustomizeStateConfigChange,  ExternalViewChange,   CustomizedViewChange,   TargetExternalViewChange,   PeriodicalRebalance,   OnDemandRebalance,     RetryRebalance,
  StateVerifier

Reply:
Actually yes, let me add these. Thanks @desaikomal

Reply:
Updated the PR. Thanks for your review. :)